### PR TITLE
RUBY-2968 Create and drop FLE2 collections

### PR DESCRIPTION
**Pull request by comandeo**

### CODE REVIEW

File: lib/mongo/operation/create/op_msg.rb [hunk 1]
```diff
@@ -31,7 +31,19 @@ class OpMsg < OpMsgBase
...
           # Collation is always supported on 3.6+ servers that would use OP_MSG.
-          spec[:selector].merge(collation: spec[:collation]).compact
+          spec[:selector].merge(
+            collation: spec[:collation],
+            encryptedFields: spec[:encrypted_fields],
+          ).compact.tap do |sel|
+            if sel[:encryptedFields] && sel[:encryptedFields].key?(:fields)
+              sel[:encryptedFields][:fields] = sel[:encryptedFields][:fields].map do |field|
+                if field[:queries] && field[:queries].key?(:contention)
+                  field[:queries][:contention] = BSON::Int64.new(field[:queries][:contention])
```
Comment (p-mongo) [inline lib/mongo/operation/create/op_msg.rb:41]: This ultimately mutates user input, doesn't it?

Comment (p-mongo) [reply lib/mongo/operation/create/op_msg.rb:41]: If `spec` is a BSON::Document then perhaps it doesn't, but I think there should be a test for this.

Comment (comandeo) [reply lib/mongo/operation/create/op_msg.rb:41]: Added the test

File: spec/mongo/operation/create/op_msg_spec.rb [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+    end
+  end
+
+  describe '#selector' do
+    min_server_fcv '3.6'
+
+    it 'does not mutate user input' do
+      user_input = spec.dup
+      described_class.new(user_input).send(:selector, connection)
+      expect(user_input).to eq(spec)
```
Comment (p-mongo) [inline spec/mongo/operation/create/op_msg_spec.rb:81]: This does not catch deep mutation. We have https://github.com/dkubb/ice_nine attached to the test suite, so  I think what would work here is

```
user_input = IceNine.deep_freeze(spec.dup)
```

and if this is modified I expect an exception to be raised.

Comment (comandeo) [reply spec/mongo/operation/create/op_msg_spec.rb:81]: Thanks, good point! Changed the test as suggested. Seems that we are fine, no exception raised.

Comment (comandeo) [reply spec/mongo/operation/create/op_msg_spec.rb]: Good point, thank you! Updated the test.

**Review by p-mongo — APPROVED** [review]

**Review by Neilshweky — APPROVED** [review]

File: lib/mongo/client.rb [hunk 3]
```diff
@@ -1141,6 +1141,14 @@ def canonicalize_ruby_options(options)
       end
     end
 
+    # Returns encrypted field map has if provided when creating the client.
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongo/client.rb]: Fixed, thanks

File: lib/mongo/collection/queryable_encryption.rb [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+            session: session,
+          )
+          do_drop(operation, session, context)
+        end
+        yield
+      end
+
+      private
+
+      # Checks is names for auxiliary collections are set and returns them,
```
[the remark this answers is not in the document]
Comment (comandeo) [reply lib/mongo/collection/queryable_encryption.rb]: Fixed, thanks

**Pull request merged**